rfcs/0001-multi-tenancy/README.md

```diff
@@ -2,37 +2,82 @@
+multi-tenancy](https://github.com/fluxcd/flux2-multi-tenancy/tree/v0.1.0), but not explained exactly
+how they relate to Flux's authorisation model. This RFC explains two multi-tenancy implementations,
+their security properties, and how they are implemented within the authorisation model.
 
 ### Goals
 
+- Explain the mechanisms available in Flux for supporting multi-tenancy
 - Define two models for multi-tenancy, "soft multi-tenancy" and "hard multi-tenancy".
 - Explain when each model is appropriate.
-- List the tenancy models supported by Flux.
 - Describe a reference implementation of each model with Flux.
 
 ### Non-Goals
 
-- Give an exhaustive account of multi-tenancy implementations.
-- Provide an [end-to-end workflow](](https://github.com/fluxcd/flux2-multi-tenancy/tree/v0.1.0))
-  of how to setup multi-tenancy with Flux.
+- Give an exhaustive account of multi-tenancy implementations in general.
+- Provide an [end-to-end workflow](](https://github.com/fluxcd/flux2-multi-tenancy/tree/v0.1.0)) of
+  how to set up multi-tenancy with Flux.
 
 ## Introduction
 
-Flux allows different organizations and/or teams to share the same Kubernetes control plane.
-Flux enables segmentation and isolation of resources across tenants by using
-Kubernetes Cluster API, namespaces and role-based access control.
+Flux allows different organizations and/or teams to share the same Kubernetes control plane; this is
+referred to as "multi-tenancy". To make this safe, Flux supports segmentation and isolation of
+resources by using namespaces and role-based access control ("RBAC"), and integrating with
+Kubernetes Cluster API.
 
-## User Roles
+The following subsections explain the existing mechanisms used for safe multi-tenancy.
+
+### Flux's authorisation model
+
+Flux defers to Kubernetes' native RBAC to specify which operations are authorised when processing
+the custom resources in the Flux API. By default, this means operations are constrained by the
+service account under which the controllers run, which (again, by default) has the `cluster-admin`
+role bound to it. This is convenient for a deployment in which all users are trusted.
+
+In a multi-tenant deployment, each tenant needs to be restricted in the operations that can be done
+on their behalf. Since tenants control Flux via its API objects, this becomes a matter of attaching
+RBAC rules to Flux API objects. There are two mechanisms that do this, "impersonation" and "remote
+apply".
+
+#### Impersonation
+
+The Kustomize controller and Helm controller both apply arbitrary sets of Kubernetes configuration
+to a cluster. These controllers are subject to authorisation on two counts:
+
+ - when accessing the Kubernetes resources that specify a particular "apply" operation -- for
+   example, the `GitRepository` object that gives access to the piece of configuration to be
+   applied; and,
+ - when creating, updating and deleting Kubernetes resources in the process of applying a piece of
```

Comment:
Also health checks, querying for secrets and configmap (valuesFrom, decryption, kubeConfig)

Reply:
That's what I meant inn the previous line item. All the things it needs to prepare for applying. Are the things that do and don't get fetched with the impersonated account documented (in one place)?

Reply:
After KC creates the Kubernetes client using the service account impersonation, it runs the following actions in order:
- fetch decryption secret
- decrypt the source
- build the Kustomize overlay
- fetch the variables values from secrets and configmaps
- run post-build substitutions
- generate the finally manifests
- validate the manifests using dry-run server-side apply
- apply the manifests using server-side apply
- wait for reconciled objects to become ready using kstatus

Reply:
I think we need to put this in the documentation, then refer to it from a summary here, rather than spelling it out here. I'll put in a PR for that.

Reply:
> After KC creates the Kubernetes client using the service account impersonation, it runs the following actions in order:
> 
>  - fetch decryption secret

This is true in the sense that the decryption secret is fetched _after_ the controller creates the impersonating account, but misleading, because the impersonating account is not used for fetching decryption secrets:
https://github.com/fluxcd/kustomize-controller/blob/main/controllers/kustomization_controller.go#L606